src/plugin_host.jl

```diff
@@ -0,0 +1,174 @@
+    else
+        proc = require_plug(h, method)
+
+        if proc == nothing
+            reply_error(c, serial, "Command not defined.")
+        end
+
+        @async(try
+            reply_result(c, serial, proc(c, args...))
+        catch
+            err = "No matching method for args (NvimClient, " * join(args, ", ") * ")"
+            reply_error(c, serial, err)
+        end)
+    end
+end
+
+function require_plug(h::HostHandler, name::String)
+    (plugin_file, proc_id) = split(name, ':', 2)
+
+    if plugin_file ∉ h.loaded_plugins
+        cbs = get_callbacks(plugin_file)
+        merge!(h.proc_callbacks, cbs)
+    end
+
+    get(h.proc_callbacks, name, nothing)
+end
+
+function get_specs(plugin_file)
+    plugin = nothing
+    try
+        plugin = decorate(plugin_file)
+    end
+
+    fn_specs = {}
+    global plug(proc_type, name, handler, opt_args...) = begin
+        conf = Dict{String, Any}()
+        conf["type"] = proc_type
+        conf["name"] = name
+        opts = Dict{String, Any}()
+        for (opt_k, opt_v) in opt_args
+            opts[string(opt_k)] = opt_v
+        end
+        conf["opts"] = opts
+        conf["sync"] = pop!(opts, "sync", 0)
+        push!(fn_specs, conf)
+    end
+
+    try
+        eval(plugin)
+    catch err
+        println(STDERR, "Error while loading plugin " * plugin_file)
+        println(STDERR, err)
+    end
+
+    fn_specs
+end
+
+function get_callbacks(plugin_file)
+    plugin = nothing
+    try
+        plugin = decorate(plugin_file)
+    end
+
+    proc_callbacks = Array((String, Expr), 0)
+    global plug(proc_type, name, handler, opt_args...) = begin
+        pattern = ""
+        for arg in opt_args
+            if arg[1] == "pattern"
+                pattern = ":" * arg[2]
+            end
+        end
+
+        proc_name = "$plugin_file:$proc_type:$name$pattern"
+        push!(proc_callbacks, (proc_name, parse(handler)))
+    end
+
+    try
+        eval(plugin)
+    catch err
+        println(STDERR, "Error while loading plugin " * plugin_file)
+        println(STDERR, err)
+    end
+
+    extract_fns(x) = (x[1], eval(x[2]))
+    Dict{String, Function}(map(extract_fns, proc_callbacks))
+end
+
+# finds constructs that look like decorators (i.e. macrocall then function)
+# and then adds the function name as the last argument to the macrocall
+function decorate(file_name::String)
+    ast = (open(readall, file_name) |> parse)::Expr
+    module_name = ast.args[2]
+
+    q = Array(Expr, 0)
+    candidate_fns = Array((Expr, Expr), 0)
+    push!(q, ast)
+    while length(q) > 0
+        expr = shift!(q)
+        for (i, arg) in enumerate(expr.args)
+            if typeof(arg) != Expr continue end
+            if arg.head == :macrocall && expr.args[i + 2].head == :function
+                # `i + 2` to skip LineNumberNode
+                fn_name = expr.args[i + 2].args[1].args[1]
+                push!(arg.args, symbol("$module_name.$fn_name"))
```

Comment:
I tried playing with a few options. I'll split my findings into macrocall syntax and then the syntax for getting the function name into the macrocall.

**macrocalls:**
```julia
1. @autocmd BufEnter, pattern="*.jl", sync=false
2. @cmd Rm(args="*")
3. @fn(Random, sync=false)
```
becomes (in the format {arg1, arg2, arg3})
```julia
1. {(BufEnter, pattern) = ("*.jl", sync) = (false,...)}
2. {(Rm(args="*"))}
3. {Random, sync=false}
```
Although Option 3 is easiest to handle, the callback name needs to be placed inside of the parens. Although slightly redundant, this allows the plugins to be defined at any location in the file.
Option 2, with a bit of massaging, works fairly well, too. It plays particularly nicely with the `->` syntax (detailed below).

**callback names**
Now,  as for capturing the function, I looks like `->` and `=`, when used with Option 2, causes the callback to be wrapped in a convenient. I prefer `=` since it's slightly more semantically correct, though I'm open to suggestions.

**summary**
I think that Option 2 with `=` (i.e. `@cmd Cmd(...) = random()`) is the way to go.

Also, I like the brevity of the single-liner in which the function name is the Nvim command name. I think that things might get confusing when Autocmds are thrown into the mix, though.